### PR TITLE
Build revamped subgraph for VoxoDeus NFT

### PR DESCRIPTION
**Pull request by hhio618**
### Deployed subgraph
- [x] Deploy on thegraph
### Description
- [x] Able to query a list of all minters efficiently
- [x] Able to query a list of all holders efficiently
- [x] Able to query a list of current holders efficiently
- [x] Able to query by user address, and this should show the unique ids of VoxoDeus NFT’s minted + held + historically owned + numbers for each
- [x] Able to query a list of top minters, top holders and top historical holders
- [x] Updated schema for more efficiency
- [x] Add required mappings to address the required queries
- [x] Add sample queries to the readme file
- [x] Add a docker-compose file for local deployments
- [x] Able to query a list of all sales related to VoxoDeus NFT on OpenSea
- [x] Able to query a list of sales by user for VoxoDeus NFT’s

### CODE REVIEW

**Comment by araa47** [comment]
Great work @hhio618 , tried deploying this on the legacy explorer but I seem to run into an error and syncing fails. 
Tried to view the logs but this is all I see on the explorer: . I also tried updating `subgraph.yaml` file to include all the new entities but run into failed syncing. 

Screenshot 2021-08-21 at 16 56 02

Some notes:

- Burn Events should be pretty uncommon but its good that you handle this too :) 
- Good work with documenting most items in the code 
- Feel free to rename the entities to something more understandable (ie: trove -> currentCollection) , realized the names might be confusing for someone not familiar with the subgraph

File: src/mapping.ts [hunk 1]
```diff
@@ -1,103 +1,128 @@
...
-// We react to only transfer events of Voxos 
-export function handleTransfer(event: Transfer): void {
-  // extract most useful fields 
-  let to = event.params.to
-  let from = event.params.from
-  let tokenId = event.params.tokenId.toI32()
-  let txid = event.transaction.hash.toHexString()
-  // assume its not a mint 
-  let mint = false
+let ZERO_ADDRESS = i32(0)
```
Comment (araa47) [inline src/mapping.ts]: Switching this to 
```
let ZERO_ADDRESS = "0x0000000000000000000000000000000000000000"
```

 solves the issue with deployment

File: src/mapping.ts [hunk 2]
```diff
@@ -1,103 +1,128 @@
...
+  
+  // Increase holdHistCount if the samaritan didn't own this trove previously.
+  if (created) {
+    samaritan.holdHistCount = samaritan.holdHistCount + 1
   }
-  // if mint add to mint 
-  if (mint){
+  
+  // Check if it's a mint or a burn event.
+  if (from.toI32() == ZERO_ADDRESS){
```
Comment (araa47) [inline src/mapping.ts]: switching this to 
```
from.toHex() == ZERO_ADDRESS
```
solves issue with deployment

File: src/mapping.ts [hunk 3]
```diff
@@ -1,103 +1,128 @@
...
-  ownageHist.push(tokenId)
-  toSamaritan.hodlHist = ownageHist
-  let trove = toSamaritan.trove
-  trove.push(tokenId)
-  toSamaritan.trove = trove 
-  stats.currentHodlers = currentHodlers
-  // save  
-  toSamaritan.save()
-  stats.save()
+  else if (to.toI32() == ZERO_ADDRESS){
```
Comment (araa47) [inline src/mapping.ts]: switching this to to.toHex() == ZERO_ADDRESS

```
from.toHex() == ZERO_ADDRESS
```
solves issue with deployment

**Comment by araa47** [comment]
@hhio618, I managed to solve the issue with deployments based on the comments above. You can try querying the subgraph  over here https://thegraph.com/legacy-explorer/subgraph/araa47/voxodeusexplorer.

**Comment by araa47** [comment]
@hhio618 Based on the example queries currently I notice a few things that are wrong.
```
{
   voxoSamaritan(id: "0xfc9648bb08a2fa852be32b2f0aaa96d32a9b7c8d"){
    id
    mintCount
    mintHist{
      tokenId
    }
    troveCount
    trove{
      id
    }
    holdHistCount
    hodlHist{
      id
    }
  }
}
```

[ ] This user should currently own 10 Voxos, however the trove count (currentCollectionCount) is inaccurate over here. The trove is however accurate. 

[ ] The hodlHist and holdHistCount are also not implemented yet. (Number of voxos owned historically + the voxo ids)

**Comment by araa47** [comment]
[ ] When querying by voxo id, the owner history is also not implemented. It should contain all the historical +  current owners of the nft

**Comment by hhio618** [comment]
Thanks for the fixes @araa47. I'm working toward fixing the problems you mentioned.

**Comment by araa47** [comment]
Hey @hhio618 in the meantime I tried deploying latest version from this PR. 

Seems like the syncing fails on the graph legacy explorer. Not really sure exactly why this time around. You can check the logs here https://thegraph.com/legacy-explorer/subgraph/araa47/voxodeusexplorer?version=pending , only has a single line output though so not very useful. 

Probably since its still work in progress, but thought Id let you know :)

**Comment by hhio618** [comment]
Looks like it's fixed now! please take a look!

**Comment by araa47** [comment]
Hey @hhio618 , im still testing but its looking good now, have deployed [here](https://thegraph.com/legacy-explorer/subgraph/araa47/voxodeusexplorer?version=current) . 

I have one more request to display the minters address when querying by Voxo Token ID.

Specifically this method 
```
{
  voxoToken(id: "1"){
    ownerHistory{
      id
    }
    user{
      id
    }
  }
}
```

It would be very useful if it kept track of the address that minted this token (this wont ever change ) , on top of showing ownerHistory + currentOwner

**Comment by hhio618** [comment]
Hey @araa47, I added the minter as you requested.
I tested the latest commit using Thegraph studio and the docker-compose, please let me know if there is any other problems.

**Comment by hhio618** [comment]
I did some digging for the bonus parts. I have some ideas, but they look a little tricky!

**Comment by araa47** [comment]
@hhio618 , managed to test out few of the methods, seems like we are good, haven't thoroughly managed to test everything yet since Ive been pretty busy. 

Regarding the bonus [this is the event](https://etherscan.io/tx/0x774e623a804922a749751b6f0e715b31845c5538aeb3256ed2c8c03ecc7fbd84#eventlog) you probably need to parse, the OrdersMatched event. However note that sales can sometime be in a different token and is not always in ethereum .

**Comment by hhio618** [comment]
Hey @araa47. It took me a lot of trial and error as there was no strict way to do the sales part, but it's finally ready to test. please take a look at it.

**Comment by araa47** [comment]
Hey @hhio618 , that's amazing, i will try to test over the weekend. I know I missed this in the requirements but it would also be useful if we could query a sale based on the NFT ID. 

So if we query VoxoDeus 1, we can get all the sales history for it.

**Comment by hhio618** [comment]
It's done! Also, I updated the readme file and add sample queries for sales data.

**Comment by hhio618** [comment]
Hey @araa47! did you had the time to check the last version?

**Comment by hhio618** [comment]
Hey, @araa47  are you okay? I was waiting to you to ping me back after the work is done!

**Pull request merged**